src/commands/protocols/common/handle-protocol-message-command.js

```diff
@@ -40,6 +45,161 @@ class HandleProtocolMessageCommand extends Command {
+            blockchain,
+            keyword,
+            await this.blockchainModuleManager.getR2(blockchain),
+            hashFunctionId,
+            false,
+        );
+        for (const { peer_id } of closestNodes) {
+            if (peer_id === this.networkModuleManager.getPeerId().toB58String()) {
+                return true;
+            }
+        }
+        this.logger.warn(`Invalid neighborhood for ual: ${ual}`);
+
+        return false;
+    }
+
+    async validateAssertionId(blockchain, contract, tokenId, assertionId, ual) {
+        const blockchainAssertionId = await this.blockchainModuleManager.getLatestAssertionId(
+            blockchain,
+            contract,
+            tokenId,
+        );
+        if (blockchainAssertionId !== assertionId) {
+            throw Error(
+                `Invalid assertion id for asset ${ual}. Received value from blockchain: ${blockchainAssertionId}, received value from request: ${assertionId}`,
+            );
+        }
+    }
+
+    async validateBid(
+        contract,
+        tokenId,
+        keyword,
+        hashFunctionId,
+        blockchain,
+        assertionId,
+        operationId,
+    ) {
+        const geAgreementData = async () => {
+            const agreementId = await this.serviceAgreementService.generateId(
+                blockchain,
+                contract,
+                tokenId,
+                keyword,
+                hashFunctionId,
+            );
+            this.logger.info(
+                `Calculated agreement id: ${agreementId} for contract: ${contract}, token id: ${tokenId}, keyword: ${keyword}, hash function id: ${hashFunctionId}, operationId: ${operationId}`,
+            );
+
+            return {
+                agreementId,
+                agreementData: await this.blockchainModuleManager.getAgreementData(
+                    blockchain,
+                    agreementId,
+                ),
+            };
+        };
+
+        const getAsk = async () => {
+            const peerRecord = await this.repositoryModuleManager.getPeerRecord(
+                this.networkModuleManager.getPeerId().toB58String(),
+                blockchain,
+            );
+
+            const ask = this.blockchainModuleManager.convertToWei(blockchain, peerRecord.ask);
+
+            return this.blockchainModuleManager.toBigNumber(blockchain, ask);
+        };
+
+        const [{ agreementId, agreementData }, blockchainAssertionSize, r0, ask] =
+            await Promise.all([
+                geAgreementData(),
+                this.blockchainModuleManager.getAssertionSize(blockchain, assertionId),
+                this.blockchainModuleManager.getR0(blockchain),
+                getAsk(),
+            ]);
+
+        const divisor = this.blockchainModuleManager
+            .toBigNumber(blockchain, r0)
+            .mul(agreementData.epochsNumber)
+            .mul(blockchainAssertionSize);
+
+        const serviceAgreementBid = this.blockchainModuleManager
```

Comment:
We should decide which formula to use here:
1. We only validate if part that is bigger than initial assertion satisfies ask (because you agreed on initial terms of current epoch): `updateTokenAmount / ((newSize - oldSize) * (epochsNumber - currentEpochIndex))  > ask`
2. We validate if full payments satisfies ask (like it's done here): `(tokenAmount + updateTokenAmount) / (newSize * (epochsNumber - currentEpochIndex)) > ask`